### PR TITLE
Improve playlist Add Songs UX and Runtime diagnostics

### PR DESCRIPTION
**Pull request by pumni**
## Summary

- Align playlist Add Songs with the playlist header, add a chevron affordance, clarify playlist add-mode copy, and add an empty-playlist Browse All Songs CTA.
- Present sender-side completion metrics with precise labels, expose production-safe pre-call lateness metrics, and plot signed recent completion residuals without fabricating unavailable distributions.
- Keep diagnostic samples session-aware, use playback lifecycle to hide finished or failed sessions, distinguish backend unavailable from no session, remove fabricated Logs, exclude high-rate snapshots/progress noise from Events, and add human-readable event timestamps.

## Verification

- cargo xtask check desktop — PASS (web checks, browser E2E, native cargo checks, and native tests).
- cargo xtask check static — PASS (security, bindings/contract, and repository static checks).
- Bun typecheck, lint, format check — PASS.
- Targeted Vitest — 30 tests passed.

## Contract notes

- Completion p50/p95/sigma and observer-only session/late/release aggregates are nullable when the active dispatch profile does not publish observer samples.
- DiagnosticsSnapshotDto exposes production-safe SendInput pre-call metrics independently, so Production can show measured pre-call values while observer-only metrics remain Unavailable.
- Release counters use 
ull for unavailable observer telemetry and

### CODE REVIEW

**Review by pumni — COMMENTED** [review]
Nghiệm thu kỹ thuật: **chưa đạt để merge** vì còn các lỗi semantic trong Runtime diagnostics trên đường chạy native production.

Điểm chính: desktop tạo `NativeDispatchSession` với `DispatchProfile::Production`. Ở profile này, engine test hiện có xác nhận `recent_latency_samples_available == false` và `recent_latencies_us` rỗng; `percentile_ms([])` lại trả 0.0. Vì vậy PR đang trình bày một số giá trị 0 như số đo completion hợp lệ dù production không thu phân phối đó. Ngoài ra lifecycle/no-session và trạng thái `unavailable` vẫn bị diễn giải sai trong các trường hợp terminal/dry-run.

CI của PR hiện đã xanh ở các gate bắt buộc, nhưng các lỗi dưới đây là lỗi mô hình dữ liệu/semantics nên test hiện tại chưa bắt được. Sau khi sửa, cần bổ sung test bám đúng native-production contract (không chỉ mock DTO có số liệu giả).

GitHub không cho tài khoản tác giả PR tự submit REQUEST_CHANGES/APPROVE, nên review này được gửi dưới dạng COMMENT; về mặt nghiệm thu tôi coi các mục inline bên dưới là blocker trước khi merge.

File: desktop/src/components/utility/DiagnosticsView.tsx [hunk 1]
```diff
@@ -65,26 +94,68 @@ export function DiagnosticsView({ useStore }: DiagnosticsViewProps) {
...
-          </p>
+          ) : !activeSession || !latest ? (
+            <DiagnosticsEmptyState
+              title="No active playback session"
+              detail="Runtime timing and input-health metrics appear when playback starts."
+            />
+          ) : (
+            <>
+              <MetricGroup title="Timing">
+                <Metric label="Completion p50" value={`${number(latest.p50_ms)} ms`} />
```
Comment (pumni) [inline desktop/src/components/utility/DiagnosticsView.tsx]: **Blocker — production không có dữ liệu completion distribution này.** Desktop native hiện tạo player với `DispatchProfile::Production`; engine test của profile này xác nhận `recent_latency_samples_available == false` và `recent_latencies_us.is_empty()`. `NativeDiagnosticsSample::from_player()` không mang cờ availability sang DTO, còn `percentile_ms([])` trả `0.0`, nên các card `Completion p50/p95/jitter` (và plot dựa trên p95) sẽ hiển thị `0.00 ms` như một số đo thật trên app production. Hãy hoặc (a) expose các metric production-safe đang có sẵn trong snapshot (`max_sendinput_pre_call_lateness_us`, `pre_call_late_*`) qua DTO, hoặc (b) truyền availability và render `Unavailable`/ẩn Timing distribution. Không nên nghiệm thu các số 0 hiện tại như telemetry đo được.

File: desktop/src/components/utility/DiagnosticsView.tsx [hunk 2]
```diff
@@ -44,8 +73,8 @@ export function DiagnosticsView({ useStore }: DiagnosticsViewProps) {
   const diagnostics = useStore((store: DesktopStore) => store.diagnostics);
   const scrollRef = useScrollVisibility<HTMLDivElement>();
   const eventsScrollRef = useScrollVisibility<HTMLDivElement>();
-  const logsScrollRef = useScrollVisibility<HTMLDivElement>();
   const latest = diagnostics.samples[diagnostics.samples.length - 1];
+  const activeSession = latest?.session_id !== null && latest?.session_id !== undefined;
```
Comment (pumni) [inline desktop/src/components/utility/DiagnosticsView.tsx]: **Lifecycle bị stale sau khi session kết thúc.** `activeSession` chỉ suy ra từ `latest.session_id`. Native monitor chỉ publish diagnostics khi còn `NativeActivePlayback`; khi terminal nó bỏ `self.active` và không phát một snapshot `session_id: null`. Vì vậy snapshot cuối của session cũ vẫn làm UI coi session là active/healthy vô thời hạn sau `finished`/`failed`. Hãy dùng `store.playback.state/sessionId` để phân biệt active vs last session, hoặc retire/clear diagnostics khi nhận terminal event (và xử lý race snapshot cuối). Bổ sung test cho `playing -> finished/failed -> Runtime` trên contract giống native.

File: desktop/src/components/utility/DiagnosticsView.tsx [hunk 3]
```diff
@@ -103,41 +174,46 @@ export function DiagnosticsView({ useStore }: DiagnosticsViewProps) {
...
+  status: DesktopStore['diagnostics']['samples'][number]['backend_status'];
+}) {
+  const label =
+    status === 'healthy'
+      ? 'Healthy'
+      : status === 'degraded'
+        ? 'Degraded'
+        : status === 'error'
+          ? 'Error'
+          : 'No session';
```
Comment (pumni) [inline desktop/src/components/utility/DiagnosticsView.tsx]: `backend_status === 'unavailable'` không đồng nghĩa với `No session`. Native có test `diagnostics_dry_run_publishes_unavailable_snapshot`: dry-run vẫn có `session_id: Some(...)` nhưng backend là `Unavailable`. Nhãn hiện tại tạo trạng thái mâu thuẫn: session đang active nhưng badge ghi `No session`. Nên dùng `Unavailable` (hoặc `Dry run / unavailable`) cho enum này; `No active playback session` phải đến từ lifecycle/session state riêng.

File: desktop/src/components/utility/DiagnosticsView.tsx [hunk 4]
```diff
@@ -11,10 +12,36 @@ function number(value: number | null | undefined, digits = 2): string {
...
+    </div>
+  );
+}
+
 function TimingPlot({ samples }: { samples: DesktopStore['diagnostics']['samples'] }) {
   const width = 560;
   const height = 112;
-  const values = samples.map((sample) => Math.max(0, sample.max_lateness_us));
+  const hasSession = samples.length > 0 && samples.at(-1)?.session_id !== null;
+  const values = hasSession ? samples.map((sample) => Math.max(0, sample.p95_ms)) : [];
```
Comment (pumni) [inline desktop/src/components/utility/DiagnosticsView.tsx]: `p95_ms` được tính từ `recent_latencies_us: Vec<i64>` nên là số signed; `Math.max(0, ...)` làm thay đổi dữ liệu. Khi p95 âm, caption có thể nói `-0.xx ms` nhưng polyline lại nằm ở 0. Nếu series này còn được giữ sau khi giải quyết availability, hãy bảo toàn sign + zero reference line, hoặc đổi nhãn rõ ràng thành một phép biến đổi positive-lateness-only thay vì gọi đây là p95 nguyên gốc.

File: desktop/src/components/utility/DiagnosticsView.tsx [hunk 5]
```diff
@@ -65,26 +94,68 @@ export function DiagnosticsView({ useStore }: DiagnosticsViewProps) {
...
+                  value={
+                    latest.release_max_us === null
+                      ? 'Unavailable'
+                      : `${number(latest.release_max_us, 0)} μs`
+                  }
+                />
+                <Metric
+                  label="Release > 2 ms"
+                  value={
+                    latest.release_late_2ms === null
```
Comment (pumni) [inline desktop/src/components/utility/DiagnosticsView.tsx]: `release_late_2ms` hiện được native map bằng `(snapshot.release_late_2ms > 0).then_some(...)`, nên `null` cũng là cách encode **zero late releases**, không chỉ 'unavailable'. Sau khi đã có release sample nhưng count = 0, UI này sẽ nói `Unavailable` thay vì `0`. Cần tách availability khỏi count (ưu tiên sửa DTO), hoặc ít nhất suy luận zero khi release timing đã có dữ liệu.

File: desktop/src/components/utility/DiagnosticsView.tsx [hunk 6]
```diff
@@ -11,41 +12,96 @@ function number(value: number | null | undefined, digits = 2): string {
...
   const eventsScrollRef = useScrollVisibility<HTMLDivElement>();
-  const logsScrollRef = useScrollVisibility<HTMLDivElement>();
   const latest = diagnostics.samples[diagnostics.samples.length - 1];
+  const playbackIsActive = ['starting', 'playing', 'paused', 'stopping'].includes(playback.state);
+  const activeSession =
+    diagnostics.enabled &&
+    playbackIsActive &&
+    playback.sessionId !== null &&
+    latest?.session_id === playback.sessionId;
+  const backendMetricsAvailable = latest?.backend_status !== 'unavailable';
```
Comment (pumni) [inline desktop/src/components/utility/DiagnosticsView.tsx]: **Blocker — `backend_status` không phải availability flag cho observer timing.** Trong engine hiện tại, `DispatchProfile::Production.observer_enabled()` là `false`; `record_lateness()` (nguồn của `max_lateness_us`, `late_2/5/10ms`, `release_max_us`, `release_late_2ms` và `recent_latencies`) chỉ được gọi từ `worker/dispatch/observer.rs`. Vì vậy production có thể có `backend_status: healthy` và pre-call metrics hợp lệ, nhưng các completion/session-max/release field observer-only vẫn chỉ là zero mặc định. `backendMetricsAvailable = backend_status !== 'unavailable'` khiến UI hiển thị các zero đó như số đo thật. Hãy thêm availability riêng cho observer/completion metrics (hoặc làm các field observer-only nullable), và chỉ dùng backend status cho backend/input-health/pre-call data.

File: desktop/src/components/utility/DiagnosticsView.test.tsx [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+      diagnostics: {
+        ...store.getState().diagnostics,
+        enabled: true,
+        samples: [
+          snapshot({
+            session_id: sessionId,
+            p50_ms: null,
+            p95_ms: null,
+            sigma_onset_ms: null,
+            backend_status: 'unavailable',
```
Comment (pumni) [inline desktop/src/components/utility/DiagnosticsView.test.tsx:111]: Test unavailable hiện dùng `backend_status: 'unavailable'`, nhưng case quan trọng của app production là **backend healthy + completion observer unavailable**. Hãy thêm fixture/test với `backend_status: 'healthy'`, `p50/p95/sigma = null`, pre-call metrics có giá trị, và xác nhận `Session max`, `Completion > ...`, `Release ...` không bị trình bày là `0` nếu chúng vẫn là observer-only. Test này sẽ khóa đúng contract mà bug còn sót ở UI hiện nay.

**Review by pumni — COMMENTED** [review]
Review vòng 2 cho commit 99f7c07: 5 blocker trước đã được xử lý đúng hướng và tôi đã resolve các thread cũ. Add Songs/lifecycle/backend unavailable/signed plot/zero release-count đều đạt.

Tuy nhiên còn **1 blocker telemetry contract** trước khi nghiệm thu merge: các field completion/session-max/release vẫn đang được coi là available chỉ dựa vào `backend_status`, trong khi trên `DispatchProfile::Production` chúng vẫn là observer-only và không được thu thập. Required CI hiện static + desktop browser đã xanh; Windows compatibility/unit tests vẫn đang chạy tại thời điểm review này.

Cần tách availability của observer completion/release metrics khỏi backend health, rồi thêm test mô phỏng đúng production contract: backend healthy + pre-call metrics available + completion distribution/observer metrics unavailable.

**Review by pumni — COMMENTED** [review]
Review vòng 3 cho commit f6ee6c9: blocker observer-only ở vòng trước đã được xử lý đúng ở contract/DTO/UI và regression test Production (backend healthy + observer unavailable) đã được bổ sung. Tôi đã resolve thread test cũ.

Còn **1 edge-case blocker nhỏ nhưng thực tế** ở trạng thái backend `unavailable` (đặc biệt dry-run): native `NativeDiagnosticsSample::unavailable()` vẫn encode các pre-call/input-health field bằng số 0 placeholder, còn UI vòng này bỏ hoàn toàn backend availability gate, nên có thể trình bày “Max pre-call lateness 0 μs” như telemetry đã đo dù không có physical/native backend. Observer metrics đã nullable đúng; cần giữ availability độc lập nhưng vẫn gate các metric production-safe/backend-dependent khi backend là unavailable.

CI commit này hiện static + desktop web/browser đã xanh; Windows compatibility/unit tests vẫn đang chạy tại thời điểm review.

File: desktop/src/components/utility/DiagnosticsView.tsx [hunk 8]
```diff
@@ -65,29 +120,70 @@ export function DiagnosticsView({ useStore }: DiagnosticsViewProps) {
...
+              detail="Runtime timing and input-health metrics appear when playback starts."
+            />
+          ) : (
+            <>
+              <MetricGroup title="Timing">
+                <Metric label="Completion p50" value={measure(latest.p50_ms, 'ms')} />
+                <Metric label="Completion p95" value={measure(latest.p95_ms, 'ms')} />
+                <Metric label="Session max" value={measure(latest.max_lateness_us, 'μs', 0)} />
+                <Metric
+                  label="Max pre-call lateness"
```
Comment (pumni) [inline desktop/src/components/utility/DiagnosticsView.tsx:153]: **Blocker nhỏ — dry-run/backend unavailable vẫn hiển thị pre-call placeholder 0 như số đo thật.** `NativeDiagnosticsSample::unavailable()` (used by the existing dry-run diagnostics test) emits `max_sendinput_pre_call_lateness_us = 0` and `pre_call_late_* = 0` because there is no physical player/backend. Sau thay đổi này UI dùng `measure/count` trực tiếp nên badge có thể là `Unavailable` nhưng card lại hiện `Max pre-call lateness 0 μs` và các pre-call counters = 0. Hãy giữ observer fields nullable như hiện tại, nhưng gate **pre-call (và tốt nhất input-health backend-dependent)** bằng `latest.backend_status !== 'unavailable'`, hoặc làm các field đó nullable trên unavailable path. Thêm test dry-run-shaped: active session + `backend_status: unavailable` + observer null => pre-call không được hiện như measured zero.

**Review by pumni — COMMENTED** [review]
Review vòng 4 cho commit b86f770: edge case dry-run/backend unavailable đã được xử lý đúng.

- Pre-call metrics được gate theo backend availability, nên dry-run không còn hiển thị placeholder 0 như telemetry đã đo.
- Input-health metrics được gate cùng contract, phù hợp vì unavailable path không có physical/native player.
- Observer-only completion/session/release metrics vẫn nullable độc lập, nên Production physical vẫn giữ được pre-call thật trong khi observer metrics là Unavailable.
- Regression test dry-run-shaped đã khóa đúng case active session + backend unavailable + zero placeholders.
- Không thấy regression mới trong diff 2 file của commit này.

Tôi coi **code review/semantic acceptance đã PASS** cho scope #210. Điều kiện còn lại trước merge chỉ là required CI của chính commit b86f770 phải xanh. Tại thời điểm review, CI #851 đã khởi chạy; classify validation đã success, còn static/security, Windows compatibility/unit tests và desktop web/browser đang in progress.

GitHub không cho tác giả PR tự APPROVE PR của mình, nên ghi nhận này là nghiệm thu kỹ thuật dưới dạng COMMENT. Khi required CI xanh toàn bộ, PR có thể merge.
